ext/standard/array.c

```diff
@@ -294,6 +294,56 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
 }
 /* }}} */
 
+static int php_zval_compare_strict_function(zval *z1, zval *z2)
+{
+	ZVAL_DEREF(z1);
+	ZVAL_DEREF(z2);
+
+	zend_uchar t1 = Z_TYPE_P(z1);
+	zend_uchar t2 = Z_TYPE_P(z2);
+	if (t1 != t2) {
+		return (t1 < t2) ? -1 : 1;
+	}
+
+	switch (t1) {
+		case IS_NULL:
+		case IS_FALSE:
+		case IS_TRUE:
+			return 0;
+		case IS_LONG:
+			return ZEND_THREEWAY_COMPARE(Z_LVAL_P(z1), Z_LVAL_P(z2));
+		case IS_DOUBLE: {
+			double d1 = Z_DVAL_P(z1);
+			double d2 = Z_DVAL_P(z2);
+			if (isnan(d1)) {
+				return -1;
+			} else if (isnan(d2)) {
+				return 1;
+			} else {
+				return ZEND_THREEWAY_COMPARE(d1, d2);
+			}
+		}
+		case IS_STRING:
+			return zend_binary_zval_strcmp(z1, z2);
```

Comment:
Separately from my previous comment: This can be optimized, the sort order is internal and not user-visible.

So if any algorithms using sorting internally (where the sort order is not user-visible) are added to php, they can avoid string comparisons entirely in the common cases (comparing hash first):

I haven't benchmarked this, though

(for the worst case of long strings with common prefixes, or the normal case of short strings with common prefixes and a mix of lengths)

- compare `ZSTR_HASH(s1)` to `ZSTR_HASH(s2)` first (compute if not already computed) - these should almost always be different for typical inputs
- Then check if they're the same pointer
- Then by length
- Then by memcmp

```c
#define ZSTR_H(zstr)    (zstr)->h
#define ZSTR_HASH(zstr) zend_string_hash_val(zstr) // gets hash value - computes the hash if it isn't already computed
```

Reply:
Good idea, let's do that!